src/common/middleware/call-api.js

```diff
@@ -50,39 +50,26 @@ export default defaults => () => next => action => {
                 }));
               }
 
-              return Promise.resolve(result);
+              // resolve promise with response JSON
+              return result;
             });
-        })
-        .catch((error) => {
-          if (failureType) {
-            next(createAction({
-              type: failureType,
-              payload: {
-                error: error
-              },
-              error: true
-            }));
-          }
-
-          error.action = action;
-
-          return Promise.reject(error);
         });
     })
     .catch((error) => {
       if (failureType) {
+        // if type of failure action is defined in settings
+        // catch the error and pass it in failure action payload
         next(createAction({
           type: failureType,
           payload: {
             error: error
           },
           error: true
         }));
+      } else {
+        // otherwise, reject the promise with error
+        throw error;
```

Comment:
So the updated `callApi` middleware always returns a promise.

In case of an error, if the error action type was specified it will dispatch the error action and **resolve** the promise (because error was passed in the action). If the error action is not specified, the promise will be **rejected** with the error.